internal/eventlog/cursor/cursor.go

```diff
@@ -0,0 +1,100 @@
+// Package cursor implements time-ordered item cursors for an event log.
+package cursor
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// A Source produces cursors based on a time index generator and a sequence
+// counter. A zero-valued Source is ready for use with defaults as described.
+type Source struct {
+	// This function is called to produce the current time index.
+	// If nil, it defaults to time.Now().UnixNano().
+	TimeIndex func() int64
+
+	// The current counter value used for sequence number generation.  It is
+	// incremented in-place each time a cursor is generated.
+	Counter int64
+}
+
+func (s *Source) timeIndex() int64 {
+	if s.TimeIndex == nil {
+		return time.Now().UnixNano()
+	}
+	return s.TimeIndex()
+}
+
+func (s *Source) nextCounter() int64 {
+	s.Counter++
+	return s.Counter
+}
+
+// Cursor produces a fresh cursor from s at the current time index and counter.
+func (s *Source) Cursor() Cursor {
+	return Cursor{
+		timestamp: uint64(s.timeIndex()),
+		sequence:  uint16(s.nextCounter() & 65535),
+	}
+}
+
+// A Cursor is a unique identifier for an item in a time-ordered event log.
+// It is safe to copy and compare cursors by value.
+type Cursor struct {
+	timestamp uint64 // ns since Unix epoch
+	sequence  uint16 // sequence number
+}
+
+// Before reports whether c is prior to o in time ordering. This comparison
+// ignores sequence numbers.
+func (c Cursor) Before(o Cursor) bool { return c.timestamp < o.timestamp }
+
+// Diff returns the time duration between c and o. The duration is negative if
+// c is before o in time order.
+func (c Cursor) Diff(o Cursor) time.Duration {
+	return time.Duration(c.timestamp) - time.Duration(o.timestamp)
+}
+
```

Comment:
But we don't have a fixed time, we have a sliding window: So we need to know not just "is x cursor after y" but "is x cursor more than d time before y" to tell whether we've exceeded the window.

I guess we could translate that, e.g., convert each cursor into a timestamp and back everytime we want to do the check. I wonder what the real advantage of that is, though? The user doesn't see the timestamp—Diff gives a duration "by whatever means is relevant".

Am I missing another advantage to the deeper abstraction here?

Reply:
One thing we _could_ do is to collapse the cursor package into the eventlog package, and unexport these methods entirely. I kinda thought clients might want to be able to compare cursors, but I don't actually have a use case for that right now. I made separate packages mostly for ease of reading at this point.

Reply:
> I thought it made sense to have Before for the case where we really want a pure order comparison

Fair, but why not actually use the full ordering of time + sequence number? One could be before the other in the total ordering despite occurring at the same instant. 

Separate packages are nice because they keep the set of tests nicely packaged together, for my money I prefer small packages that do one thing.

Reply:
> Fair, but why not actually use the full ordering of time + sequence number? One could be before the other in the total ordering despite occurring at the same instant. 

There are two questions a cursor answers: one is the identity of a particular event item, for which the sequence number is just a disambiguator (if we knew the clock always has enough resolution we could maybe just trust that. But we can't be sure the system clock will always give us ns even though that's the units—so we keep a little nonce the clock can't control). For this purpose we use equality ("this is me"). Here sequence does matter, but not the order of sequences. Just "is it the same one at this moment".

The other concern is relative order, which we use for maintaining the queue. For that identity doesn't matter, and comparing the sequence numbers is harmless but also useless. The user doesn't care, it's purely an implementation detail. As long as earlier in time is detectable. 

The user of the package doesn't really care, so mainly this is a matter of the simplest viable implementation. We could constrain it more, but would that buy us anything?

> Separate packages are nice because they keep the set of tests nicely packaged together, for my money I prefer small packages that do one thing.

Agree. 👍🏼

Reply:
Handling time in the way you describe would definitely work. That said: I would prefer not to add another abstraction ("timestamp") that the user of a Cursor would have to keep track of explicitly. The two concerns the caller cares about don't require it.

We could always extend the API surface later if we wind up needing to. For now I'd prefer to keep it fairly direct (i.e., addressing the semantics, rather than exposing the internals).